src/DocumentDB.ChangeFeedProcessor/Utils/TaskExtensions.cs

```diff
@@ -28,5 +28,23 @@ public static async Task IgnoreException(this Task task)
                 // ignore
             }
         }
+
+        public static async Task<TResult> AbortOnTimeout<TResult>(this Task<TResult> responseTask, TimeSpan timeout, CancellationToken token)
+        {
+            using (CancellationTokenSource cancelSource = CancellationTokenSource.CreateLinkedTokenSource(token))
+            {
+                Task timeoutTask = Task.Delay(timeout, cancelSource.Token);
+
+                if (responseTask == await Task.WhenAny(responseTask, timeoutTask).ConfigureAwait(false))
+                {
+                    cancelSource.Cancel(); // cancel the delay task
+                    return responseTask.GetAwaiter().GetResult(); // propagate exception
```

Comment:
that would throw if original task was cancelled, right? Is that the reason to use GetAwaiter?

Reply:
this code is executed when the original task is not cancelled but completed or failed with exception. Then we cancel the timeout task and call GetAwaiter to propagate possible exception of the original task

Reply:
Task.WhenAny will break from waiting when any of the tasks is cancelled as well. The code looks fine though.